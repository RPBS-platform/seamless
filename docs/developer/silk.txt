See plans/silk.md for a decent description of Silk.
Not everything in there has been implemented, but the essentials are there.