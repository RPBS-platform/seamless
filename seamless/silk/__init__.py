from .Silk import Silk, is_none
from jsonschema.exceptions import FormatError, ValidationError
