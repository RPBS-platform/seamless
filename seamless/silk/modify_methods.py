from .validation import _array_types
from functools import partial

def list_grow_method(self, name, *args, **kwargs):
    item = None
    item_schema = None
    if name == "append":
        item, = args
        pos = len(self)
    elif name == "insert":
        pos, item = args
    if item is not None:
        if isinstance(item, Silk):
            item_schema = item.schema.dict

    method = self._get_special(name, skip_modify_methods = True)

    if isinstance(item, Silk):
        item_schema = item.schema.dict
        item = item.data

    schema = self._schema
    schema_updated = self._infer_new_item(
       schema, pos, item, value_item_schema=item_schema
    )

    result = method(*args, **kwargs)
    if not len(self._forks):
        self.validate()
    if schema_updated and self._schema_update_hook is not None:
        self._schema_update_hook()
    return result

def list_modify_method(self, name, *args, **kwargs):
    #TODO: special case for numpy arrays, including resize() if needed,
    #   but need a path from the parent!
    method = self._get_special(name, skip_modify_methods = True)
    result = method(*args, **kwargs)
    if not len(self._forks):
        self.validate()

def dict_modify_method(self, name, *args, **kwargs):
    #TODO: special case for numpy arrays, including astype() if needed,
    #   but need a path from the parent!
    method = self._get_special(name, skip_modify_methods = True)
    result = method(*args, **kwargs)
    if not len(self._forks):
        self.validate()

_list_grow_method_names = set(("append", "extend", "insert"))
_list_modify_method_names = set(("clear", "pop", "remove", "reverse", "sort"))
_dict_modify_method_names = set(("clear", "pop", "popitem", "update"))

def try_modify_methods(self, method):
    """Tries if "method" is a data-modifying method
    Returns:
        is_modifying_method: bool
        result: partial method
    """
    if isinstance(self.data, _array_types):
        if method in _list_grow_method_names:
            result = partial(list_grow_method, self, method)
            return True, result
        elif method in _list_modify_method_names:
            result = partial(list_modify_method, self, method)
            return True, result
        else:
            return False, None
    elif isinstance(self.data, dict):
        if method in _dict_modify_method_names:
            result = partial(dict_modify_method, self, method)
            return True, result
        else:
            return False, None
    else:
        return False, None

from .Silk import Silk
