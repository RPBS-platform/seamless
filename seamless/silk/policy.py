default_policy = {
    "surplus_to_binary": "error",
    "accept_missing_binary_optional": True,
    "infer_property": True,
    "infer_item":  "uniform",
    "infer_type":  True,
    "infer_shapedarray":  True,
    "infer_dtype":  False,
    "infer_dtype_mixed":  False,
    "infer_default": False,
    "error_log": True,
    "infer_recursive":  False,
    "binary_validation": True,
    "infer_shape":  False,
    "re_infer_shape":  False,
    "infer_required": False,
    "wrap_scalar": False
}

#TODO: code facility to globally override default_policy
