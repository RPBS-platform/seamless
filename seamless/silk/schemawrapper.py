import weakref
from pprint import pformat
from copy import deepcopy

from .. import Wrapper
class SchemaWrapper(Wrapper):
    _update_hook = None
    _parent = lambda self: None
    def __init__(self, _parent, _dict, _update_hook):
        if _parent is not None:
            super().__setattr__("_parent", weakref.ref(_parent))
        if isinstance(_dict, SchemaWrapper):
            _dict = _dict.dict
        super().__setattr__("_dict", _dict)
        if _update_hook is not None:
            super().__setattr__("_update_hook", _update_hook)

    def _unwrap(self):
        return self._dict

    def _get(self, attribute):
        child = self._dict.get(attribute, None)
        if child is None:
            props = self._dict.get("properties")
            if props is None:
                raise AttributeError(attribute)
            child = props.get(attribute)
            if child is None:
                raise AttributeError(attribute)
        return SchemaWrapper(self._parent(), child, self._update_hook)

    def copy(self):
        return SchemaWrapper(None, deepcopy(self._dict), None)

    def update(self, value):
        if isinstance(value, SchemaWrapper):
            value = value.dict
        self._dict.update(value)
        parent = self._parent()
        if parent is not None:
            parent.validate(accept_none=True)
        if self._update_hook is not None:
            self._update_hook()

    def pop(self, attribute):
        child = self._dict.get(attribute, None)
        if child is None:
            props = self._dict.get("properties", None)
            if props is None:
                raise AttributeError(attribute)
            child = props.get(attribute, None)
            if child is None:
                raise AttributeError(attribute)
            result = props.pop(attribute)
        else:
            result = self._dict.pop(attribute)
        parent = self._parent()
        if parent is not None:
            parent.validate(accept_none=True)
        if self._update_hook is not None:
            self._update_hook()
        return result

    def __delattr__(self, attribute):
        self.pop(attribute)

    def __setattr__(self, attribute, value):
        if isinstance(value, SchemaWrapper):
            value = value.dict
        self._dict[attribute] = value
        parent = self._parent()
        if parent is not None:
            parent.validate(accept_none=True)
        if self._update_hook is not None:
            self._update_hook()

    def __setitem__(self, item, value):
        setattr(self, item, value)

    def __getattribute__(self, attribute):
        if attribute == "dict": #TODO: property with docstring
            return super().__getattribute__("_dict")
        if attribute in ("pop", "copy", "update"):
            return super().__getattribute__(attribute)
        if isinstance(attribute, str) and attribute.startswith("_"):
            return super().__getattribute__(attribute)
        return self._get(attribute)

    def __getitem__(self, item):
        return getattr(self, item)

    def __repr__(self):
        return pformat(self._dict)

    def __str__(self):
        return pformat(self._dict)
