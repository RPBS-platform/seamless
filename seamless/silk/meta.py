from .schemawrapper import SchemaWrapper
from .Silk import Silk
import json

def meta(name, bases, d):
    # For now, ignore __module__ and __qualname__
    d.pop("__module__")
    d.pop("__qualname__")
    schema = d.pop("schema").dict
    s = Silk()
    s.schema.dict.update(schema)
    prototype = {}
    with s.fork():
        for key, value in d.items():
            if isinstance(value, Validator):
                s._add_validator(value.func, attr=None, from_meta=True, name=None)
            elif callable(value) or isinstance(value, property):
                setattr(s, key, value)
            else:
                try:
                    json.dumps(value)
                except Exception:
                    raise ValueError("'%s' is not JSON-serializable" % key)
                prototype[key] = value
        if len(prototype):            
            s.schema["__prototype__"] = prototype

    return Silk(schema=s.schema.dict)

def prep(name, bases):
    return {
        "schema": SchemaWrapper(None, {}, None),
    }

meta.__prepare__ = prep


class Validator:
    def __init__(self, func):
        self.func = func

def validator(func):
    return Validator(func)
