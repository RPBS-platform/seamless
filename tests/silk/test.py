import numpy as np

class Test:
    z = [1,2,3]
    def __len__(self):
        return len(self.z)
    def __iter__(self):
        return iter(self.z)
    def __getitem__(self, item):
        print("ITEM", item)
        return self.z[item]

t = Test()
a = np.array(t)
print(a, type(a), a.shape)


from silk import Silk
s = Silk().set([10,20,30])
aa = np.array(s)
print(aa, type(aa), aa.shape)
print(s[:2])
