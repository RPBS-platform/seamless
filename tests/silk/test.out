]0;IPython: silk/test[('origin', [('x', <class 'numpy.float32'>), ('y', <class 'numpy.float32'>), ('z', <class 'numpy.float32'>)]), ('x', [('x', <class 'numpy.float32'>), ('y', <class 'numpy.float32'>), ('z', <class 'numpy.float32'>)]), ('y', [('x', <class 'numpy.float32'>), ('y', <class 'numpy.float32'>), ('z', <class 'numpy.float32'>)]), ('z', [('x', <class 'numpy.float32'>), ('y', <class 'numpy.float32'>), ('z', <class 'numpy.float32'>)])]
((10.0, 0.0, 0.0), (0.0, 0.0, 0.0), (0.0, 0.0, 0.0), (0.0, 0.0, 0.0))
class _Coordinate_method_block_1:
        __slots__ = []
    
        def __mul__(self, a):
            """Multiplication with either:
              a Float (scaling)
              a Coordinate (dot p
def _validate(self):
    z = self.z
    y = self.y
    x = self.x
    def _validation_block_1():
        
            try:
                assert abs(x**2+y**2+z**2 - 1) < 0.001
            except Exception as exc:
                raise SilkValidationError("The size of the vector must be 1, not {'%.3f' % abs(x**2+y**2+z**2)}", globals(), locals())
          
    _validation_block_1()
{'x': 1.0, 'y': 2.0, 'z': 3.0}
Vector (
  x = 0.6,
  y = 0.8,
  z = 0.0,
)
(1.0, 2.0, 3.0)
(1.0, 2.0, 20.0)
CoordinateArray (
  Coordinate (
    x = 1.0,
    y = 2.0,
    z = 3.0,
  ),
  Coordinate (
    x = 4.0,
    y = 5.0,
    z = 6.0,
  ),
  Coordinate (
    x = 7.0,
    y = 8.0,
    z = 9.0,
  ),
)
CoordinateArray (
  Coordinate (
    x = 1.0,
    y = 2.0,
    z = 3.0,
  ),
  Coordinate (
    x = 4.0,
    y = 5.0,
    z = 6.0,
  ),
  Coordinate (
    x = 7.0,
    y = 8.0,
    z = 9.0,
  ),
)
(array([(1.0, 2.0, 3.0), (4.0, 5.0, 6.0), (7.0, 8.0, 9.0), (0.0, 0.0, 0.0),
       (0.0, 0.0, 0.0), (0.0, 0.0, 0.0), (0.0, 0.0, 0.0), (0.0, 0.0, 0.0),
       (0.0, 0.0, 0.0), (0.0, 0.0, 0.0)], 
      dtype=[('x', '<f4'), ('y', '<f4'), ('z', '<f4')]), array([3], dtype=uint16))
CoordinateArray (
  Coordinate (
    x = 1.0,
    y = 2.0,
    z = 3.0,
  ),
  Coordinate (
    x = 4.0,
    y = 5.0,
    z = 6.0,
  ),
  Coordinate (
    x = 7.0,
    y = 8.0,
    z = 9.0,
  ),
)
CoordinateArrayArray (
  CoordinateArray (
    Coordinate (
      x = 1.0,
      y = 2.0,
      z = 3.0,
    ),
    Coordinate (
      x = 4.0,
      y = 5.0,
      z = 6.0,
    ),
    Coordinate (
      x = 7.0,
      y = 8.0,
      z = 9.0,
    ),
  ),
  CoordinateArray (
    Coordinate (
      x = 10.0,
      y = 20.0,
      z = 30.0,
    ),
    Coordinate (
      x = 40.0,
      y = 50.0,
      z = 60.0,
    ),
    Coordinate (
      x = 70.0,
      y = 80.0,
      z = 90.0,
    ),
    Coordinate (
      x = 0.0,
      y = 1.0,
      z = 2.0,
    ),
  ),
)
CoordinateArrayArray (
  CoordinateArray (
    Coordinate (
      x = 1.0,
      y = 2.0,
      z = 3.0,
    ),
    Coordinate (
      x = 4.0,
      y = 5.0,
      z = 6.0,
    ),
    Coordinate (
      x = 7.0,
      y = 8.0,
      z = 9.0,
    ),
  ),
  CoordinateArray (
    Coordinate (
      x = 10.0,
      y = 20.0,
      z = 30.0,
    ),
    Coordinate (
      x = 40.0,
      y = 50.0,
      z = 60.0,
    ),
    Coordinate (
      x = 70.0,
      y = 80.0,
      z = 90.0,
    ),
    Coordinate (
      x = 0.0,
      y = 1.0,
      z = 2.0,
    ),
  ),
) [2 3 4]
CoordinateArrayArray (
  CoordinateArray (
    Coordinate (
      x = 1.0,
      y = 2.0,
      z = 3.0,
    ),
    Coordinate (
      x = 4.0,
      y = 5.0,
      z = 6.0,
    ),
    Coordinate (
      x = 7.0,
      y = 8.0,
      z = 9.0,
    ),
  ),
  CoordinateArray (
    Coordinate (
      x = 10.0,
      y = 20.0,
      z = 30.0,
    ),
    Coordinate (
      x = 40.0,
      y = 50.0,
      z = 60.0,
    ),
    Coordinate (
      x = 70.0,
      y = 80.0,
      z = 90.0,
    ),
    Coordinate (
      x = 0.0,
      y = 1.0,
      z = 2.0,
    ),
  ),
)
CoordinateArrayArrayArray (
  CoordinateArrayArray (
    CoordinateArray (
      Coordinate (
        x = 1.0,
        y = 2.0,
        z = 3.0,
      ),
      Coordinate (
        x = 4.0,
        y = 5.0,
        z = 6.0,
      ),
      Coordinate (
        x = 7.0,
        y = 8.0,
        z = 9.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = 10.0,
        y = 20.0,
        z = 30.0,
      ),
      Coordinate (
        x = 40.0,
        y = 50.0,
        z = 60.0,
      ),
      Coordinate (
        x = 70.0,
        y = 80.0,
        z = 90.0,
      ),
      Coordinate (
        x = 0.0,
        y = 1.0,
        z = 2.0,
      ),
    ),
  ),
  CoordinateArrayArray (
    CoordinateArray (
      Coordinate (
        x = -1.0,
        y = -2.0,
        z = -3.0,
      ),
      Coordinate (
        x = -4.0,
        y = -5.0,
        z = -6.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = -10.0,
        y = -20.0,
        z = -30.0,
      ),
      Coordinate (
        x = -40.0,
        y = -50.0,
        z = -60.0,
      ),
      Coordinate (
        x = -70.0,
        y = -80.0,
        z = -90.0,
      ),
      Coordinate (
        x = 0.0,
        y = -1.0,
        z = -2.0,
      ),
      Coordinate (
        x = -12.0,
        y = -11.0,
        z = -10.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = -4.0,
        y = -5.0,
        z = -6.0,
      ),
      Coordinate (
        x = -7.0,
        y = -8.0,
        z = -9.0,
      ),
    ),
  ),
  CoordinateArrayArray (
    CoordinateArray (
      Coordinate (
        x = 1.0,
        y = 2.0,
        z = 3.0,
      ),
      Coordinate (
        x = 4.0,
        y = 5.0,
        z = 6.0,
      ),
      Coordinate (
        x = 7.0,
        y = 8.0,
        z = 9.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = 10.0,
        y = 20.0,
        z = 30.0,
      ),
      Coordinate (
        x = 40.0,
        y = 50.0,
        z = 60.0,
      ),
      Coordinate (
        x = 70.0,
        y = 80.0,
        z = 90.0,
      ),
      Coordinate (
        x = 0.0,
        y = 1.0,
        z = 1002.0,
      ),
    ),
  ),
)
START
CoordinateArrayArrayArray (
  CoordinateArrayArray (
    CoordinateArray (
      Coordinate (
        x = 1.0,
        y = 2.0,
        z = 3.0,
      ),
      Coordinate (
        x = 4.0,
        y = 5.0,
        z = 6.0,
      ),
      Coordinate (
        x = 7.0,
        y = 8.0,
        z = 9.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = 10.0,
        y = 20.0,
        z = 30.0,
      ),
      Coordinate (
        x = 40.0,
        y = 50.0,
        z = 60.0,
      ),
      Coordinate (
        x = 70.0,
        y = 80.0,
        z = 90.0,
      ),
      Coordinate (
        x = 0.0,
        y = 1.0,
        z = 2.0,
      ),
    ),
  ),
  CoordinateArrayArray (
    CoordinateArray (
      Coordinate (
        x = -1.0,
        y = -2.0,
        z = -3.0,
      ),
      Coordinate (
        x = -4.0,
        y = -5.0,
        z = -6.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = -10.0,
        y = -20.0,
        z = -30.0,
      ),
      Coordinate (
        x = -40.0,
        y = -50.0,
        z = -60.0,
      ),
      Coordinate (
        x = -70.0,
        y = -80.0,
        z = -90.0,
      ),
      Coordinate (
        x = 0.0,
        y = -1.0,
        z = -2.0,
      ),
      Coordinate (
        x = -12.0,
        y = -11.0,
        z = -10.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = -4.0,
        y = -5.0,
        z = -6.0,
      ),
      Coordinate (
        x = -7.0,
        y = -8.0,
        z = -9.0,
      ),
    ),
  ),
  CoordinateArrayArray (
    CoordinateArray (
      Coordinate (
        x = 1.0,
        y = 2.0,
        z = 3.0,
      ),
      Coordinate (
        x = 4.0,
        y = 5.0,
        z = 6.0,
      ),
      Coordinate (
        x = 7.0,
        y = 8.0,
        z = 9.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = 10.0,
        y = 20.0,
        z = 30.0,
      ),
      Coordinate (
        x = 40.0,
        y = 50.0,
        z = 60.0,
      ),
      Coordinate (
        x = 70.0,
        y = 80.0,
        z = 90.0,
      ),
      Coordinate (
        x = 0.0,
        y = 1.0,
        z = 1002.0,
      ),
    ),
  ),
)
True
CoordinateArrayArrayArray (
  CoordinateArrayArray (
    CoordinateArray (
      Coordinate (
        x = 1.0,
        y = 2.0,
        z = 3.0,
      ),
      Coordinate (
        x = 4.0,
        y = 5.0,
        z = 6.0,
      ),
      Coordinate (
        x = 7.0,
        y = 8.0,
        z = 9.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = 10.0,
        y = 20.0,
        z = 30.0,
      ),
      Coordinate (
        x = 40.0,
        y = 50.0,
        z = 60.0,
      ),
      Coordinate (
        x = 70.0,
        y = 80.0,
        z = 90.0,
      ),
      Coordinate (
        x = 0.0,
        y = 1.0,
        z = 2.0,
      ),
    ),
  ),
  CoordinateArrayArray (
    CoordinateArray (
      Coordinate (
        x = -1.0,
        y = -2.0,
        z = -3.0,
      ),
      Coordinate (
        x = -4.0,
        y = -5.0,
        z = -6.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = -10.0,
        y = -20.0,
        z = -30.0,
      ),
      Coordinate (
        x = -40.0,
        y = -50.0,
        z = -60.0,
      ),
      Coordinate (
        x = -70.0,
        y = -80.0,
        z = -90.0,
      ),
      Coordinate (
        x = 0.0,
        y = -1.0,
        z = -2.0,
      ),
      Coordinate (
        x = -12.0,
        y = -11.0,
        z = -10.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = -4.0,
        y = -5.0,
        z = -6.0,
      ),
      Coordinate (
        x = -7.0,
        y = -8.0,
        z = -9.0,
      ),
    ),
  ),
  CoordinateArrayArray (
    CoordinateArray (
      Coordinate (
        x = 1.0,
        y = 2.0,
        z = 3.0,
      ),
      Coordinate (
        x = 4.0,
        y = 5.0,
        z = 6.0,
      ),
      Coordinate (
        x = 7.0,
        y = 8.0,
        z = 9.0,
      ),
    ),
    CoordinateArray (
      Coordinate (
        x = 10.0,
        y = 20.0,
        z = 30.0,
      ),
      Coordinate (
        x = 40.0,
        y = 50.0,
        z = 60.0,
      ),
      Coordinate (
        x = 70.0,
        y = 80.0,
        z = 90.0,
      ),
      Coordinate (
        x = 0.0,
        y = 1.0,
        z = 1002.0,
      ),
    ),
  ),
)
True
True
True
True
[3 2 3 4 0 3 2 5 2 2 3 4 0] (3, 3, 5)
START2
True
False
[2 2 3 4 0 2 3 4 0 0 0 0 0] (3, 3, 5) [0 0 0 0]
[3 2 3 4 0 3 2 5 2 2 3 4 0] (3, 3, 5)
False
(10, 5, 7)
[3 2 3 4 0 0 0 3 2 5 2 0 0 2 3 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] (10, 5, 7)
False
True
False
{'origin': {'z': 0.0, 'y': 0.0, 'x': 0.0}, 'z': {'z': 1.0, 'y': 0.0, 'x': 0.0}, 'y': {'z': 0.0, 'y': 1.0, 'x': 0.0}, 'x': {'z': 20.0, 'y': 2.0, 'x': 1.0}}
{'z': 0.0, 'y': 0.0, 'x': 0.0}
AxisSystem (
  origin = Coordinate (
    x = 0.0,
    y = 0.0,
    z = 0.0,
  ),
  x = Coordinate (
    x = 1.0,
    y = 2.0,
    z = 20.0,
  ),
  y = Coordinate (
    x = 0.0,
    y = 1.0,
    z = 0.0,
  ),
  z = Coordinate (
    x = 0.0,
    y = 0.0,
    z = 1.0,
  ),
)
AxisSystem (
  origin = Coordinate (
    x = 0.0,
    y = 0.0,
    z = 0.0,
  ),
  x = Coordinate (
    x = 1.0,
    y = 2.0,
    z = 20.0,
  ),
  y = Coordinate (
    x = 0.0,
    y = 1.0,
    z = 0.0,
  ),
  z = Coordinate (
    x = 0.0,
    y = 0.0,
    z = 1.0,
  ),
) <class 'seamless.silk.registers.typenames.AxisSystem'>
{'origin': {'z': 0.0, 'y': 0.0, 'x': 0.0}, 'z': {'z': 1.0, 'y': 0.0, 'x': 0.0}, 'y': {'z': 0.0, 'y': 1.0, 'x': 0.0}, 'x': {'z': 20.0, 'y': 2.0, 'x': 1.0}} <class 'dict'>
AxisSystem (
  origin = Coordinate (
    x = 0.0,
    y = 0.0,
    z = 0.0,
  ),
  x = Coordinate (
    x = 1.0,
    y = 2.0,
    z = 20.0,
  ),
  y = Coordinate (
    x = 0.0,
    y = 1.0,
    z = 0.0,
  ),
  z = Coordinate (
    x = 0.0,
    y = 0.0,
    z = 1.0,
  ),
)
{'origin': {'z': 0.0, 'y': 0.0, 'x': 0.0}, 'z': {'z': 1.0, 'y': 0.0, 'x': 0.0}, 'y': {'z': 0.0, 'y': 1.0, 'x': 0.0}, 'x': {'z': 20.0, 'y': 2.0, 'x': 1.0}}
True
((999.0, 0.0, 0.0), (1.0, 2.0, 20.0), (0.0, 1.0, 0.0), (0.0, 0.0, 1.0))
AxisSystem (
  origin = Coordinate (
    x = 999.0,
    y = 0.0,
    z = 0.0,
  ),
  x = Coordinate (
    x = 1.0,
    y = 2.0,
    z = 20.0,
  ),
  y = Coordinate (
    x = 0.0,
    y = 1.0,
    z = 0.0,
  ),
  z = Coordinate (
    x = 0.0,
    y = 0.0,
    z = 1.0,
  ),
)
(2, 2)
(2,) (2,)
1 (2,)
((1.0, 2.0), (b'three', False), b'', [(1.0, 2.0, 3.0), (10.0, 20.0, 30.0)], 29773760, [2], ((0.0, 0.0, 0.0), (0.0, 0.0, 0.0), (0.0, 0.0, 0.0), (0.0, 0.0, 0.0)), False, False)
START
Test (
  x = (
    b = 1.0,
    a = 2.0,
  ),
  y = (
    c = "three",
    d = False,
  ),
  q = CoordinateArray (
    Coordinate (
      x = -20.0,
      y = -40.0,
      z = -60.0,
    ),
    Coordinate (
      x = 1.0,
      y = 2.0,
      z = 3.0,
    ),
    Coordinate (
      x = 10.0,
      y = 20.0,
      z = 30.0,
    ),
  ),
)
True
{'x': {'b': 1.0, 'a': 2.0}, 'y': {'c': 'three', 'd': 0}, 'q': [{'x': -20.0, 'z': -60.0, 'y': -40.0}, {'x': 1.0, 'z': 3.0, 'y': 2.0}, {'x': 10.0, 'z': 30.0, 'y': 20.0}]}
{'q': [{'z': -60.0, 'y': -40.0, 'x': -20.0}, {'z': 3.0, 'y': 2.0, 'x': 1.0}, {'z': 30.0, 'y': 20.0, 'x': 10.0}], 'y': {'c': 'three', 'd': 0}, 'x': {'b': 1.0, 'a': 2.0}}
True
START2
Test (
  x = (
    b = 1.0,
    a = 2.0,
  ),
  y = (
    c = "three",
    d = True,
  ),
  q = CoordinateArray (
    Coordinate (
      x = -20.0,
      y = -40.0,
      z = -60.0,
    ),
    Coordinate (
      x = 1.0,
      y = 2.0,
      z = 3.0,
    ),
    Coordinate (
      x = 10.0,
      y = 20.0,
      z = 30.0,
    ),
  ),
)
(
  test = TestArray (
    Test (
      x = (
        b = 1.0,
        a = 2.0,
      ),
      y = (
        c = "three",
        d = False,
      ),
      q = CoordinateArray (
        Coordinate (
          x = -20.0,
          y = -40.0,
          z = -60.0,
        ),
        Coordinate (
          x = 1.0,
          y = 2.0,
          z = 3.0,
        ),
        Coordinate (
          x = 10.0,
          y = 20.0,
          z = 30.0,
        ),
      ),
    ),
    Test (
      x = (
        b = -2.0,
        a = -1.0,
      ),
      y = (
        c = "three",
        d = False,
      ),
      q = CoordinateArray (
        Coordinate (
          x = -20.0,
          y = -40.0,
          z = -60.0,
        ),
        Coordinate (
          x = 1.0,
          y = 2.0,
          z = 3.0,
        ),
        Coordinate (
          x = 10.0,
          y = 20.0,
          z = 30.0,
        ),
      ),
    ),
  ),
)
{'q': [{'z': -60.0, 'y': -40.0, 'x': -20.0}, {'z': 3.0, 'y': 2.0, 'x': 1.0}, {'z': 30.0, 'y': 20.0, 'x': 10.0}], 'z': None, 'y': {'c': 'three', 'd': 0}, 'ax': {'origin': {}, 'z': {}, 'y': {}, 'x': {}}, 'x': {'b': -2.0, 'a': -1.0}}
Test (
  x = (
    b = -2.0,
    a = -1.0,
  ),
  y = (
    c = "three",
    d = False,
  ),
  q = CoordinateArray (
    Coordinate (
      x = -20.0,
      y = -40.0,
      z = -60.0,
    ),
    Coordinate (
      x = 1.0,
      y = 2.0,
      z = 3.0,
    ),
    Coordinate (
      x = 10.0,
      y = 20.0,
      z = 30.0,
    ),
  ),
)
JSON
TestArray (
  Test (
    x = (
      b = 1.0,
      a = 2.0,
    ),
    y = (
      c = "three",
      d = True,
    ),
    q = CoordinateArray (
      Coordinate (
        x = -20.0,
        y = -40.0,
        z = -60.0,
      ),
      Coordinate (
        x = 1.0,
        y = 2.0,
        z = 3.0,
      ),
      Coordinate (
        x = 10.0,
        y = 20.0,
        z = 30.0,
      ),
    ),
  ),
  Test (
    x = (
      b = -2.0,
      a = -1.0,
    ),
    y = (
      c = "three",
      d = True,
    ),
    q = CoordinateArray (
      Coordinate (
        x = -20.0,
        y = -40.0,
        z = -60.0,
      ),
      Coordinate (
        x = 1.0,
        y = 2.0,
        z = 3.0,
      ),
      Coordinate (
        x = 10.0,
        y = 20.0,
        z = 30.0,
      ),
    ),
  ),
)
NUMPY
[[3]
 [3]] [3]
[[3]
 [4]] [4]
TestArray (
  Test (
    x = (
      b = 1.0,
      a = 2.0,
    ),
    y = (
      c = "three",
      d = False,
    ),
    q = CoordinateArray (
      Coordinate (
        x = -20.0,
        y = -40.0,
        z = -60.0,
      ),
      Coordinate (
        x = 1.0,
        y = 2.0,
        z = 3.0,
      ),
      Coordinate (
        x = 10.0,
        y = 20.0,
        z = 30.0,
      ),
    ),
  ),
  Test (
    x = (
      b = -2.0,
      a = -1.0,
    ),
    y = (
      c = "three",
      d = False,
    ),
    q = CoordinateArray (
      Coordinate (
        x = -20.0,
        y = -40.0,
        z = -60.0,
      ),
      Coordinate (
        x = 1.0,
        y = 2.0,
        z = 3.0,
      ),
      Coordinate (
        x = 10.0,
        y = 20.0,
        z = 30.0,
      ),
      Coordinate (
        x = 7.0,
        y = 7.0,
        z = 7.0,
      ),
    ),
    ax = AxisSystem (
      origin = Coordinate (
        x = 4.0,
        y = 7.0,
        z = 8.0,
      ),
      x = Coordinate (
        x = 2.0,
        y = 2.0,
        z = 2.0,
      ),
      y = Coordinate (
        x = 6.0,
        y = 2.0,
        z = 0.0,
      ),
      z = Coordinate (
        x = 0.0,
        y = 0.0,
        z = 1.0,
      ),
    ),
  ),
)
numpy numpy
True
True
